parse orca compound jobs

joegair opened the pull request:
**overall:** 
rcaFile.read_file('path') will now parse compound orca files and return a list of OrcaFile objects.

**In parse_orca.py:**
1. telling cctk to look for regex "COMPOUND JOB  \d{1,}" to indicate breaks in compound jobs
2. the first output_block of a compound job does not contain new info about the calculation. so if there is >1 output_block, the first output_block is ommitted.

**In orca_file.py**
1. change the way that charge and multiplicity are parsed for compatibility with compound job. (not compatible with miniprint keyword)
2. change the way successful_termination and elapsed_time is counted for compatibility with compound job. 
    -. I'm concerned that this approach is super inefficient and the relevant info might already be extracted elsewhere.
    -. Currently, it only counts successful completion of SP, OPT, FREQ, and NMR jobs at this time)

Next challenge comes from the fact that job_types are defined with class job_types(enum), so each job_type has a name:value pair. This works well in g16, where all optimizations contain the keyword "opt", but this leads to a problem in ORCA when there are several keywords that can describe the same job_type, eg tightopt, normalopt, copt, zopt,  etc.

ekwan commented:
Can you also update the documentation (maybe the ORCA recipe?) so that we know what kind of job verbosity is needed for ORCA?  I am not sure what you mean by name:value pair in job_types.  Enums are basically lists, not dicts.  Can you just call all those opts an optimization, or does it make more sense to have more enums?

joegair commented:
> Can you also update the documentation (maybe the ORCA recipe?) so that we know what kind of job verbosity is needed for ORCA? I am not sure what you mean by name:value pair in job_types. Enums are basically lists, not dicts. Can you just call all those opts an optimization, or does it make more sense to have more enums?

I updated the docs and the unit tests to reflect the problem with MiniPrint. Several additional changes to orca_file and parse_orca to support getting frequencies from compound jobs and getting the right frequencies when calc_hess is used in a TS search.

ekwan commented:
Thanks.  Here are some comments:

- documentation: I see the comment about MiniPrint--thanks for adding.  Would you please move ORCA to a new recipe and provide more details about what kinds of files can be read (or written)?  Please provide examples of all common use cases. Please also explain how multi-step jobs can be parsed (or written).
- OrcaJobType: Could you please explain the new enum members?  Note that each enum should have its own doctag.  What's the difference between TS, OPTTS, SCANTS, and TIGHTOPT?  Do we need separate enums for all of these?  Why are there no expected properties for the new enum members?  Shouldn't we be reading all the gradients and all that for all these new job types?
- Please don't comment out the old code (replace it instead).  I note this occurred on parse_orca/read_freqs, for example.  Git will store the old code.
- Are there unit tests that test these new features?

joegair commented:
Hi Eugene,

I cleaned up the OrcaJobType(Enum) using tuples to match the values to expected properties.

I removed all of the unnecessary JobTypes.  There is one ORCA keyword (ScanTS) that is a member of OPT, but does not include "opt" in the keyword. I addressed this by specifying in the _assign_job_types method that if "ScanTs" is in the header, to add OPT to job_types.

I am working on unit tests for parsing frequencies, especially in compound jobs and jobs that involve multiple steps like ScanTS.

I edited some of the documentation to get rid of some errors that arise in "hello world" and "tutorial 1". The errors stem from the old use of file.molecule instead of file.get_molecule()

ekwan commented on cctk/orca_file.py:
```diff
@@ -197,20 +185,26 @@ def read_file(cls, filename):
+                try:
+                    properties[-1]["enthalpy"] = enthalpies[-1]
+                except Exception as e:
+                    print(f"error finding enthalpy for {filename}")
```
When would an enthalpy not be found?  Are there cases where you would have more than one enthalpy?

ekwan commented on cctk/orca_file.py:
```diff
@@ -126,7 +114,7 @@ def read_file(cls, filename):
                     successful_opt += 1
                 elif line.startswith("Maximum memory used throughout the entire SCFHESS-calculation:"): #### a frequency job was completed
-                    successful_freq += 1
+                    successful_freq += 1                                                                #### this approach has a liability, it will return a successful frequency job even if the freq job was only the initial hessian used for the TS search
```
I think it's OK to report the initial frequency job for a TS job in the count.  That's actually what happens and you can deal with that with logic on the job type.

ekwan commented on cctk/orca_file.py:
```diff
@@ -17,41 +17,29 @@ class OrcaJobType(Enum):
     """
 
-    NMR = "nmr"
+    NMR = ("nmr", ["isotropic_shielding"])
```
Please also parse the paramagnetic, diamagnetic, and total shielding tensors as n_atoms x 3 x 3 array.  Please read the anisotropy for each atom as well.

ekwan commented on cctk/orca_file.py:
```diff
@@ -17,41 +17,29 @@ class OrcaJobType(Enum):
     """
 
-    SP = "sp"
+    SP = ("sp", ["energy", "scf_iterations"])
```
Could you also extract the timings?  Or you must be doing it already for the benchmarks?

ekwan commented on cctk/ensemble.py:
```diff
@@ -113,7 +113,7 @@ def molecule_list(self):
     def properties_list(self):
         """
-        Returns a list of the constituent molecules.
+        Returns a list of dictionaries. One dictionary per geometry. Each dictionary contains the property names and property values for each geometry in the ensemble. 
```
This looks to have been a copy/paste bug in the original.  Good job catching it.

ekwan commented on docs/recipe_09.rst:
```diff
@@ -0,0 +1,314 @@
+    :align: center
+
+
+""""""""""""""""""""""""""""""""
```
No need to include these here as they are already in the repo.  You could just add links.

ekwan commented on docs/recipe_09.rst:
```diff
@@ -0,0 +1,314 @@
+.. _recipe_09:
+
+======================================
+Reading and Writing ORCA Files
```
This new recipe is really nice!

ekwan commented on docs/recipe_09.rst:
```diff
@@ -0,0 +1,314 @@
+
+- ``import cctk`` is assumed.
+- *cctk* was originally designed with Gaussian in mind, but supports basic writing of ORCA input files and parsing of ORCA output
+- Note that the MiniPrint option may make Orca output files unreadable by cctk
```
Can you be specific and say what level of printing you recommend?  Then you can say you do NOT recommend using MiniPrint.  Are we checking the print level and warning or raising an error?

ekwan commented on docs/recipe_09.rst:
```diff
@@ -0,0 +1,314 @@
+- ``import cctk`` is assumed.
+- *cctk* was originally designed with Gaussian in mind, but supports basic writing of ORCA input files and parsing of ORCA output
+- Note that the MiniPrint option may make Orca output files unreadable by cctk
+- All of the input and output files used in these recipes can be accessed at https://github.com/ekwan/cctk/tree/master/test/static
```
You can just leave a hyperlink instead of showing the full address.

ekwan commented on docs/recipe_09.rst:
```diff
@@ -0,0 +1,314 @@
+- *cctk* was originally designed with Gaussian in mind, but supports basic writing of ORCA input files and parsing of ORCA output
+- Note that the MiniPrint option may make Orca output files unreadable by cctk
+- All of the input and output files used in these recipes can be accessed at https://github.com/ekwan/cctk/tree/master/test/static
+- All of input files used in these recipes can also be copied as text at the end of this recipe.
```
Just give a link (see below).

ekwan commented on docs/recipe_09.rst:
```diff
@@ -0,0 +1,314 @@
+
+
+- In this recipe, we convert an ``.xyz`` file into an ORCA ``.inp`` file.
+- cctk can only write single-geometry input files for ORCA
```
Ah, so we can't do multiple "link1" style jobs?

ekwan commented on docs/recipe_09.rst:
```diff
@@ -0,0 +1,314 @@
+
+  * xyz 0 1
+  7         -2.59192634   -2.21198726    0.00000000
+  1         -1.59236634   -2.24165320    0.00000000
```
I'd prefer not to have the recipes be too long, so I would just truncate most of this output with "..."

ekwan commented on docs/recipe_09.rst:
```diff
@@ -0,0 +1,314 @@
+  *
+
+""""""""""""""""""""""""""""""""""""""""""""""""
+Writing an ORCA input file from a SMILES string
```
This is a nice example!

ekwan commented on docs/recipe_09.rst:
```diff
@@ -0,0 +1,314 @@
+    6          3.81472230    0.33750522    0.14614943
+    6          2.70351934    0.51809502   -0.57053703
+    7          1.44507718    0.26401237   -0.05548295
+    6          1.27698672   -0.21741368    1.23580658
```
... as before

ekwan commented on docs/recipe_09.rst:
```diff
@@ -0,0 +1,314 @@
+
+  path = 'test/static/orca_uridine_opt_freq.out'
+  file = cctk.OrcaFile.read_file(path)
+
```
The most common things people might want are (a) the last geometry and (b) the last energy.  What's the easiest way to get those?  Does get_molecule() work here?

ekwan commented on docs/recipe_09.rst:
```diff
@@ -0,0 +1,314 @@
+  path = 'test/static/orca_uridine_opt_freq.out'
+  file = cctk.OrcaFile.read_file(path)
+
+  file.ensemble.properties_list()
```
This is confusing because it's not actually the list of properties, it's actually a list of properties_dict objects!  I don't think we can change the API, though.  I would recommend that you show how to get the last properties_dict first, because that's the simplest case.  You can explain the key:value mapping for that case.  Then, you can explain how to get the other properties_dicts.

ekwan commented on docs/recipe_09.rst:
```diff
@@ -0,0 +1,314 @@
+
+Returns the properties dictionary of the first geometry in the ensemble::
+
+  # {'energy': -911.078699594417,
```
Change to last geometry and move up

ekwan commented on docs/recipe_09.rst:
```diff
@@ -0,0 +1,314 @@
+  freqs = file.ensemble[-1,'frequencies']
+  # assigns a list of frequencies to the variable freqs
+
+To access a given property for each member of the ensemble::
```
This is good, it's more complex and should come later in the discussion.

ekwan commented on docs/recipe_09.rst:
```diff
@@ -0,0 +1,314 @@
+
+  import matplotlib.pyplot as plt
+
+  f1 = plt.figure(figsize=(8,6))
```
might be helpful to label these as energy_figure and rms_gradient_figure

ekwan commented on docs/tutorial_01.rst:
```diff
@@ -17,14 +17,17 @@ Overview
 Many computational chemistry papers report structures in the ``.xyz`` format, which is not recognized by Gaussian. 
-Although manual conversion is facile for one file, an automated solution can prove useful for bulk data processing. 
+Although manual conversion is facile for one file, an automated solution can prove useful for bulk data processing.
+A sample ``.xyz`` file can be found at the end of this tutorial. 
```
Link, please do not paste files into recipes

ekwan commented on docs/tutorial_01.rst:
```diff
@@ -35,19 +38,20 @@ This will open a blank file. First, we need to load *cctk*::
 The above code creates a *cctk* ``XYZFile`` object from the file we specified, which now exists as a Python data structure. 
 To output a different filetype, we need to extract the ``Molecule`` object represented by the file and write it as a ``.gjf`` file::
 
+    molecule = file.get_molecule()
```
Does this work for ORCA now?

joegair replied:
yes, added to recipe 9

ekwan commented:
Nice work!  I see things are coming along!

ekwan commented:
Also, please don't forget to add orca tests.

ekwan commented:
New action items: check for charge parsing (mulliken, lowdin, hirshfeld, etc), check that charges are properly connected to the geometry iteration, check ORCA version, ensure free energy parsing works between ORCA 4 and 5

ekwan commented:
Great work!  Merged as requested.
